crates/bevy_ecs/src/schedule/stepping.rs

```diff
@@ -0,0 +1,1557 @@
+            None => return None,
+            Some(label) => label,
+        };
+        let state = match self.schedule_states.get(label) {
+            None => return None,
+            Some(state) => state,
+        };
+        state
+            .node_ids
+            .get(self.cursor.system)
+            .map(|node_id| (*label, *node_id))
+    }
+
+    /// Enable stepping for the provided schedule
+    pub fn add_schedule(&mut self, schedule: impl ScheduleLabel) -> &mut Self {
+        self.updates.push(Update::AddSchedule(schedule.intern()));
+        self
+    }
+
+    /// Disable stepping for the provided schedule
+    ///
+    /// NOTE: This function will also clear any system-specific behaviors that
+    /// may have been configured.
+    pub fn remove_schedule(&mut self, schedule: impl ScheduleLabel) -> &mut Self {
+        self.updates.push(Update::RemoveSchedule(schedule.intern()));
+        self
+    }
+
+    /// Clear behavior set for all systems in the provided [`Schedule`]
+    pub fn clear_schedule(&mut self, schedule: impl ScheduleLabel) -> &mut Self {
+        self.updates.push(Update::ClearSchedule(schedule.intern()));
+        self
+    }
+
+    /// Begin stepping at the start of the next frame
+    pub fn enable(&mut self) -> &mut Self {
+        #[cfg(feature = "bevy_debug_stepping")]
+        self.updates.push(Update::SetAction(Action::Waiting));
+        #[cfg(not(feature = "bevy_debug_stepping"))]
+        error!(
+            "Stepping cannot be enabled; \
+            bevy was compiled without the bevy_debug_stepping feature"
+        );
+        self
+    }
+
+    /// Disable stepping, resume normal systems execution
+    pub fn disable(&mut self) -> &mut Self {
+        self.updates.push(Update::SetAction(Action::RunAll));
+        self
+    }
+
+    /// Check if stepping is enabled
+    pub fn is_enabled(&self) -> bool {
+        self.action != Action::RunAll
+    }
+
+    /// Run the next system during the next render frame
+    ///
+    /// NOTE: This will have no impact unless stepping has been enabled
+    pub fn step_frame(&mut self) -> &mut Self {
+        self.updates.push(Update::SetAction(Action::Step));
+        self
+    }
+
+    /// Run all remaining systems in the stepping frame during the next render
+    /// frame
+    ///
+    /// NOTE: This will have no impact unless stepping has been enabled
+    pub fn continue_frame(&mut self) -> &mut Self {
+        self.updates.push(Update::SetAction(Action::Continue));
+        self
+    }
+
+    /// Ensure this system always runs when stepping is enabled
+    ///
+    /// Note: if the system is run multiple times in the [`Schedule`], this
+    /// will apply for all instances of the system.
+    pub fn always_run<Marker>(
+        &mut self,
+        schedule: impl ScheduleLabel,
+        system: impl IntoSystem<(), (), Marker>,
+    ) -> &mut Self {
+        let type_id = IntoSystem::into_system(system).type_id();
```

Comment:
Ideally we don't need to actually construct the system to get the TypeId here. We should be able to rely on `TypeId::of::<IntoSys::System>()`. However currently that is out of sync with `FunctionSystem`'s `type_id()` implementation. I'll add a `PERF` todo.

Reply:
Definitely not worth blocking on.